### DISCUSSION TITLE
pyo3-asyncio: Is with_bound(...).unbind() an antipattern?

### DISCUSSION THREAD

**Discussion by mhils** [PyO3/pyo3#4299] [General]
Hi folks! With @davidhewitt's fantastic effort to get a pyo3_asyncio_0_21 fork rolling, we've upgraded to the new Bound API. This was all smooth sailing, but we have one pattern that looks a bit funny:

Old code:
```rust
    /// async def read(n: int) -> bytes:
    fn read<'py>(&self, py: Python<'py>, n: u32) -> PyResult<&'py PyAny> {
        // ...
        pyo3_asyncio::tokio::future_into_py(py, async move {
            let data: Vec<u8> = rx.await.unwrap();
            Python::with_gil(|py| Ok(PyBytes::new(py, &data).into_py(py)))
        }
    }
```

New code:
```rust
    fn read<'py>(&self, py: Python<'py>, n: u32) -> PyResult<Bound<'py, PyAny>> {
        // ...
        pyo3_asyncio_0_21::tokio::future_into_py(py, async move {
            let data: Vec<u8> = rx.await.unwrap();
            Python::with_gil(|py| Ok(PyBytes::new_bound(py, &data).unbind()))
        }
    }
```

Note how we now do `PyBytes::new_bound(py, &data).unbind()` - we first bind, but then we immediately unbind again. That feels wrong somehow. Are we holding things wrong, is there a better/more correct way?

**Comment by davidhewitt** [comment]
Great question and sorry for the slow reply.

As I guess you're aware, the async interaction here adds a lot of complexity because the GIL can't be held across the `await` yield points. So it's expected that there will be a lot of switching back and forth between `Py<T>` for storage and `Bound<'py, T>` for interactions.

As such, I think `PyBytes::new_bound(py, &data).unbind()` is expected here. It's only the case that the `PyBytes` constructor returns a `Bound<'py, PyBytes>` rather than `Py<PyBytes>` because we already had to pass the `Python<'py>` token in, and we might as well carry that information in the return value for convenience.

----

In PyO3 0.23 we will be completing the removal of the `gil-refs` API and with that comes the opportunity to reclaim `PyBytes::new` for the "bound" API. So I think the canonical thing to write will soon be:

```rust
Python::with_gil(|py| Ok(PyBytes::new(py, &data).unbind()))
```

which I think is even more fine :)

**Reply by mhils** [reply]
> In PyO3 0.23 we will be completing the removal of the `gil-refs` API and with that comes the opportunity to reclaim `PyBytes::new` for the "bound" API. So I think the canonical thing to write will soon be:

This sounds fantastic! Thank you for the detailed reply, this more than answers my question. Looking forward to the new API! 😃
